test/filestorage.js

```diff
@@ -0,0 +1,262 @@
+      fs.readdirSync(dir).forEach(file => cleanup(path.join(dir, file)));
+      fs.rmdirSync(dir);
+    } else {
+      fs.unlinkSync(dir);
+    }
+  }
+};
+
+const fsTest = metatests.test('Filestorage');
+fsTest.endAfterSubtests();
+
+fsTest.beforeEach((test, cb) => {
+  create({ dir: testDir, minCompressSize }, (err, storage) => {
+    test.error(err);
+    cb({ storage });
+  });
+});
+
+fsTest.afterEach((test, cb) => {
+  cleanup(testDir);
+  cb();
+});
+
+fsTest.test('Write string to file', (test, { storage }) => {
+  storage.write(new common.Uint64(0), 'data0',
+    { checksum: 'CRC32', dedupHash: 'SHA256' },
+    (err, checksum, dedupHash, originalSize) => {
+      test.error(err);
+      test.strictSame(checksum, '20cd1c30');
+      test.strictSame(dedupHash,
+        '4285d10832a8edf4cc7979e9a257e145dc5c934549f62e0bf1dc6070e67cc4ab');
+      test.strictSame(originalSize, 5);
+
+      fs.readFile(utils.getFilepath(testDir, '0000', '0000'), (err, data) => {
+        test.error(err);
+        test.strictSame(data.toString(), 'data0');
+        test.end();
+      });
+    });
+});
+
+fsTest.test('Write Buffer to file', (test, { storage }) => {
+  storage.write(new common.Uint64(1), Buffer.from('data1'),
+    { checksum: 'CRC32', dedupHash: 'SHA256' },
+    (err, checksum, dedupHash, originalSize) => {
+      test.error(err);
+      test.strictSame(checksum, '57ca2ca6');
+      test.strictSame(dedupHash,
+        '5b41362bc82b7f3d56edc5a306db22105707d01ff4819e26faef9724a2d406c9');
+      test.strictSame(originalSize, 5);
+
+      fs.readFile(utils.getFilepath(testDir, '0001', '0000'), (err, data) => {
+        test.error(err);
+        test.strictSame(data.toString(), 'data1');
+        test.end();
+      });
+    });
+});
+
+fsTest.test('Read file', (test, { storage }) => {
+  fs.mkdir(path.join(testDir, '0002'), err => {
+    test.error(err);
+    fs.writeFile(utils.getFilepath(testDir, '0002', '0000'), 'data2', err => {
+      test.error(err);
+      storage.read(new common.Uint64(2), (err, data) => {
+        test.error(err);
+        test.strictSame(data.toString(), 'data2');
+        test.end();
+      });
+    });
+  });
+});
+
+fsTest.test('Remove file', (test, { storage }) => {
+  fs.mkdir(path.join(testDir, '0003'), err => {
+    test.error(err);
+    const file = utils.getFilepath(testDir, '0003', '0000');
+    fs.writeFile(file, 'data3', err => {
+      test.error(err);
+      storage.rm(new common.Uint64(3), err => {
+        test.error(err);
+        test.strictSame(fs.existsSync(file), false);
+        test.end();
+      });
+    });
+  });
+});
+
+fsTest.test('Compress small file using zip', (test, { storage }) => {
+  const file = utils.getFilepath(testDir, '0004', '0000');
+  fs.mkdir(path.join(testDir, '0004'), err => {
+    test.error(err);
+    fs.writeFile(utils.getFilepath(testDir, '0004', '0000'), 'data4', err => {
+      test.error(err);
+      storage.compress(new common.Uint64(4), 'ZIP', (err, compressed) => {
```

Comment:
Nit: it'd be better to have a file-local function to get unique IDs to avoid having to remember which one comes next.
i.e.
```js
let idCounter = 1;
const nextId = () => new common.Uint64(idCounter++);
```

Reply:
Actually, there is no need to track ids, because folder for testing is being cleaned before each test. It's just to quickly determine how many tests there are.